x-pack/plugins/ingest_manager/server/plugin.ts

```diff
@@ -0,0 +1,114 @@
+import { EncryptedSavedObjectsPluginStart } from '../../encrypted_saved_objects/server';
+import { SecurityPluginSetup } from '../../security/server';
+import { PluginSetupContract as FeaturesPluginSetup } from '../../features/server';
+import { PLUGIN_ID } from './constants';
+import { appContextService } from './services';
+import { registerDatasourceRoutes, registerAgentConfigRoutes } from './routes';
+import { IngestManagerConfigType } from '../common';
+
+export interface IngestManagerSetupDeps {
+  licensing: LicensingPluginSetup;
+  security?: SecurityPluginSetup;
+  features?: FeaturesPluginSetup;
+}
+
+export interface IngestManagerAppContext {
+  clusterClient: IClusterClient;
+  encryptedSavedObjects: EncryptedSavedObjectsPluginStart;
+  security?: SecurityPluginSetup;
+  license$?: Observable<ILicense>;
+  config$?: Observable<IngestManagerConfigType>;
+}
+
+export class IngestManagerPlugin implements Plugin {
+  private license$!: Observable<ILicense>;
+  private config$!: Observable<IngestManagerConfigType>;
+  private clusterClient!: IClusterClient;
+  private security: SecurityPluginSetup | undefined;
+
+  constructor(private readonly initializerContext: PluginInitializerContext) {
+    this.config$ = this.initializerContext.config.create<IngestManagerConfigType>();
+  }
+
+  public async setup(core: CoreSetup, deps: IngestManagerSetupDeps) {
+    this.license$ = deps.licensing.license$;
+    this.clusterClient = core.elasticsearch.dataClient;
+    this.security = deps.security;
+
+    // Register feature
+    // TODO: Flesh out privileges
+    if (deps.features) {
+      deps.features.registerFeature({
+        id: PLUGIN_ID,
+        name: 'Ingest Manager',
+        icon: 'savedObjectsApp',
+        navLinkId: PLUGIN_ID,
+        app: [PLUGIN_ID, 'kibana'],
+        privileges: {
+          all: {
+            api: [PLUGIN_ID],
+            savedObject: {
+              all: [],
+              read: [],
+            },
+            ui: ['show'],
+          },
+          read: {
+            api: [PLUGIN_ID],
+            savedObject: {
+              all: [],
+              read: [],
+            },
+            ui: ['show'],
+          },
+        },
+      });
+    }
+
+    // Create router
+    const router = core.http.createRouter();
+
+    // Register routes
+    registerAgentConfigRoutes(router);
+    registerDatasourceRoutes(router);
+
+    // Optional route registration depending on Kibana config
+    // restore when EPM & Fleet features are added
+    // const config = await this.config$.pipe(first()).toPromise();
+    // if (config.epm.enabled) registerEPMRoutes(router);
+    // if (config.fleet.enabled) registerFleetSetupRoutes(router);
+  }
+
+  public async start(
+    core: CoreStart,
+    plugins: {
+      encryptedSavedObjects: EncryptedSavedObjectsPluginStart;
+    }
+  ) {
+    appContextService.start({
+      clusterClient: this.clusterClient,
```

Comment:
@joshdover  - apologies I feel like I've asked you this before... isn't this an anti-pattern - holding on to state returned from coreSetup to be used outside it? The flow of state is hard to follow here, I think these things could be retrieved from RouteHandlerContext?

Reply:
We had some discussion about this in https://github.com/elastic/kibana/pull/56567#discussion_r377358655